---
title: "Filtering and picrust out"
output: html_document
date: "2025-11-25"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r library loading}
library(tidyverse)
library(dplyr)
library(ggpicrust2)
library(DESeq2)
library(GGally)
library(phyloseq)
library(microbiome)
library(ggVennDiagram)
library(vegan)
library(indicspecies)
library(writexl)
library(ggpubr) # load packages for stat_compare_means
library(dunn.test)
library(pairwiseAdonis)
```


```{r filtering and saving new filtered datsets}
# Alcohol dataset
alc_original = read.delim("../Datasets/p8_metadata.txt",  header = TRUE)
alc_no_species = alc_original %>%
  select(-Host)

alc_no_species$alcohol_intake_status = ifelse(
  (alc_no_species$host_sex == "female" & alc_no_species$GramPerday >= 25) |
    (alc_no_species$host_sex == "male" & alc_no_species$GramPerday >= 45),
  "High", "Low")

alc_no_species$age_group = ifelse(
  alc_no_species$Host_age >= 60, "Senior",
  ifelse(alc_no_species$Host_age >= 20 & alc_no_species$Host_age < 60, "Adult", "Child")
)
alc_grouped = alc_no_species
alc_removed_row_5 = alc_grouped[-5, ]
alc_removed_row_5$GramPerday = as.numeric(alc_removed_row_5$GramPerday)
alc_removed_row_5$Host_age = as.numeric(alc_removed_row_5$Host_age)
alc_fixed = alc_removed_row_5

alc = alc_fixed |>
  filter(alcohol_intake_status == 'Low')
# IBD dataset
ryan_ibd_data = read.delim('../Datasets/ryan_metadata.tsv', header = TRUE)
ryan_ibd_data_filtered = ryan_ibd_data |>
  filter(Smoking.status == 'Non-smoker') |>
  filter(Medications == 'No' | Medications == "") |>
  filter(Age_when_sampled != "Uncertain")

ryan_ibd_data_filtered$Medications = ifelse(
  ryan_ibd_data_filtered$Medications == "",  # test for empty string
  "No",                                       # value if TRUE
  ryan_ibd_data_filtered$Medications         # value if FALSE
)

# join the filtered datasets with the manifest datasets
# Alcohol
# Make sure both are data frames
alc_fixed = as.data.frame(alc_fixed)
p8_manifest = read.table("../Datasets/p8_manifest.tsv", header = TRUE, sep = "\t")
alc_merged_data = left_join(alc_fixed, p8_manifest)

# IBD
ryan_ibd_data_filtered = as.data.frame(ryan_ibd_data_filtered)
ryan_manifest = read.table("../Datasets/ryan_manifest.tsv", header = TRUE, sep = "\t")
ryan_ibd_merged_data = left_join(ryan_ibd_data_filtered, ryan_manifest)

# Combine both IBD conditions into IBD and leave in the same column Condition (Only IBD vs Healthy) for PICRUST only


# saving datasets into tsv format
# alcohol
write.table(
  alc_merged_data,                # your merged data frame
  file = "../Datasets/alc_merged_data.tsv",   # output file name
  sep = "\t",                 # tab-separated
  row.names = FALSE,          # don’t write row numbers
  quote = FALSE               # don’t quote strings unless needed
)
# IBD
write.table(
  ryan_ibd_merged_data,                # your merged data frame
  file = "../Datasets/ryan_ibd_merged_data.tsv",   # output file name
  sep = "\t",                 # tab-separated
  row.names = FALSE,          # don’t write row numbers
  quote = FALSE               # don’t quote strings unless needed
)
```

```{r pathways datasets loading}
pathways_alcohol = read.delim("../Datasets/p8alcohol_path_abun_unstrat.tsv", row.name = 1) # reading in the tsv with rownames as the first column

pathways_ibd = read.delim("../Datasets/ryan_path_abun_unstrat.tsv", row.name = 1) # reading in the tsv with rownames as the first column

```

```{r}
# Alcohol
alcohol_daa_results_df = pathway_daa(abundance = pathways_alcohol,
                              metadata = alc_merged_data,
                              group = "alcohol_intake_status",
                              daa_method = "DESeq2",
                              select = NULL, reference = NULL)

alcohol_daa_annotated_results_df = pathway_annotation(pathway = "MetaCyc",
                                               daa_results_df = alcohol_daa_results_df)

# IBD

# Merging IBD CD and UC Conditions
picrust_ibd_merged_data = ryan_ibd_merged_data %>% 
  mutate(Condition_merged = case_when(
    str_detect(Condition, "Healthy") ~ "Healthy",
    str_detect(Condition, "Crohn's Disease") ~ "IBD",
    str_detect(Condition, "Ulcerative Colitis") ~ "IBD"
  ))

ibd_daa_results_df = pathway_daa(abundance = pathways_ibd,
                              metadata = picrust_ibd_merged_data,
                              group = "Condition_merged",
                              daa_method = "DESeq2",
                              select = NULL, reference = NULL)

ibd_daa_annotated_results_df = pathway_annotation(pathway = "MetaCyc",
                                               daa_results_df = ibd_daa_results_df)
```

```{r plotting}
source('ggpicrust2_errorbar_function_fixed.R')


alcohol_daa_annotated_results_df %>% filter(p_adjust < 0.05) %>% nrow()

# Explain this!
# This makes sense, as there are no significant differences between different alcohol inflammatory environments' high level pathways 
```

```{r plotting}
# Figure 3 Using the special dataset
source('ggpicrust2_errorbar_function_fixed.R')

ibd_daa_annotated_results_df %>% filter(p_adjust < 0.05) %>% nrow()

ibd_peb = pathway_errorbar_fixed(abundance = pathways_ibd, 
                     daa_results_df = ibd_daa_annotated_results_df,
                     Group = picrust_ibd_merged_data$Condition_merged,
                     wrap_label = T, wraplength=60,
                     fc_cutoff = 0.2, order_by_log = F,
                     p_values_threshold = 0.05, # Changed the p value threshold to include at least 2 results
                     ko_to_kegg = FALSE, 
                     p_value_bar = TRUE, 
                     x_lab = "description")

ibd_peb

ggsave("PICRUStAnalysisIBD.png",
       plot = ibd_peb, width = 18, height = 10)
```

```{r core microbiome}
# Using the uncombined IBD dataset
# Loading datasets
alcohol_taxonomy = read.delim('../Datasets/p8alcohol_taxonomy.tsv', row.names = 1)
alcohol_counts = read.delim('../Datasets/p8alcohol_feature-table.txt', skip=1, row.names=1)
alc_metadata = read.delim("../Datasets/alc_merged_data.tsv", row.names = 1)

ibd_taxonomy = read.delim('../Datasets/ryan_taxonomy.tsv', row.names = 1) 
ibd_counts = read.delim('../Datasets/ryan_feature-table.txt', skip=1, row.names=1)
ibd_metadata = read.delim("../Datasets/ryan_ibd_merged_data.tsv", row.names = 1)

# Figure 2
# Alcohol

# counts datatset
alcohol_counts = as.matrix(alcohol_counts)
alc_counts = otu_table(alcohol_counts, taxa_are_rows = T)

# metadata
alc_sam_data = sample_data(alc_metadata)

# taxanomy dataset
alc_tax_wrangled = alcohol_taxonomy |>
  separate(col = Taxon, 
           into = c('Domain','Phylum','Class','Order',
                    'Family','Genus','Species'),
           sep=';', fill='right') |>
           select(-Confidence) |>
           as.matrix()
alc_tax_wrangled = tax_table(alc_tax_wrangled)

# combine
alc_ps = phyloseq(alc_counts, alc_sam_data, alc_tax_wrangled)

# print out the ps
alc_ps


# IBD

# counts datatset
ibd_counts = as.matrix(ibd_counts)
ibd_counts = otu_table(ibd_counts, taxa_are_rows = T)

# metadata
ibd_sam_data = sample_data(ibd_metadata)

# taxanomy dataset
ibd_tax_wrangled = ibd_taxonomy |>
  separate(col = Taxon, 
           into = c('Domain','Phylum','Class','Order',
                    'Family','Genus','Species'),
           sep=';', fill='right') |>
           select(-Confidence) |>
           as.matrix()
ibd_tax_wrangled = tax_table(ibd_tax_wrangled)

# combine
ibd_ps = phyloseq(ibd_counts, ibd_sam_data, ibd_tax_wrangled)

# print out the ps
ibd_ps

# Core microbiome plots

# Alcohol

# Rarefy if desired and convert to relative abundance, use tax_glom (optional)
alc_psrare = rarefy_even_depth(alc_ps, sample.size = 51, rngseed=123)
alc_ps_rare_relab = transform(alc_psrare, 'compositional')
alc_ps_rare_relab_genus = tax_glom(alc_ps_rare_relab, 'Genus')

# Subset your phyloseq object
low_alcohol_intake = subset_samples(alc_ps_rare_relab_genus, alcohol_intake_status == 'Low')
high_alcohol_intake = subset_samples(alc_ps_rare_relab_genus, alcohol_intake_status == 'High')

# Find core members
ASVs_low_alcohol_intake = core_members(low_alcohol_intake, detection=0.001, prevalence = 0.1)
ASVs_high_alcohol_intake = core_members(high_alcohol_intake, detection=0.001, prevalence = 0.1)

alcohol_venn_core_microbiome = ggVennDiagram(
    list(ASVs_low_alcohol_intake, ASVs_high_alcohol_intake),
    set_size = 5,
    category.names = c('low_alcohol','high_alcohol')) +
    theme(plot.margin = margin(8, 40, 8, 40)) +
    coord_cartesian(clip = "off") +
    scale_fill_gradientn(colors = c(low = "#C4D8F3", high = "#FFCC9E"))
alcohol_venn_core_microbiome

# Need to save this in the Figures folder inside of the Results folder

# IBD

# Rarefy if desired and convert to relative abundance, use tax_glom (optional)
ibd_psrare = rarefy_even_depth(ibd_ps, sample.size = 120, rngseed=123)
ibd_ps_rare_relab = transform(ibd_psrare, 'compositional')
ibd_ps_rare_relab_genus = tax_glom(ibd_ps_rare_relab, 'Genus')

# Subset your phyloseq object
healthy = subset_samples(ibd_ps_rare_relab_genus, Condition == 'Healthy')
crohn_disease = subset_samples(ibd_ps_rare_relab_genus, Condition == "Crohn's Disease")
ulcerative_colitis = subset_samples(ibd_ps_rare_relab_genus, Condition == "Ulcerative Colitis")

# Find core members
ASVs_healthy = core_members(healthy, detection=0.001, prevalence = 0.1)
ASVs_crohn_disease = core_members(crohn_disease, detection=0.001, prevalence = 0.1)
ASVs_ulcerative_colitis = core_members(ulcerative_colitis, detection=0.001, prevalence = 0.1)

ibd_venn_core_microbiome = ggVennDiagram(
    list(ASVs_healthy, ASVs_crohn_disease, ASVs_ulcerative_colitis),
    set_size = 4,
    category.names = c("Healthy", "CD", "UC")) +
    scale_fill_gradientn(colors = c(low = "#99D48B", high = "#D19FC0"))
ibd_venn_core_microbiome
# Need to save this in the Figures folder inside of the Results folder

# Maybe save this in the Figures folder inside of the Results folder
# ggsave(
#   filename = "ibd_venn_core_microbiome.jpeg",
#   plot = ibd_venn_core_microbiome,
#   path = "../R_scripts",                
#   width = 10,
#   height = 10)
```

```

# R Load data

```{r data}
# data for both alcohol & IBD
metadata_alcohol = read.delim("../Datasets/alc_merged_data.tsv", row.names = 1)
taxonomy_alcohol = read.delim("../Datasets/p8alcohol_taxonomy.tsv", row.names = 1)
counts_alcohol = read.delim("../Datasets/p8alcohol_feature-table.txt", skip=1, row.names=1)

metadata_ibd = read.delim("../Datasets/ryan_ibd_merged_data.tsv", row.names = 1)
taxonomy_ibd = read.delim("../Datasets/ryan_taxonomy.tsv", row.names = 1)
counts_ibd = read.delim("../Datasets/ryan_feature-table.txt", skip = 1, row.names = 1)

```

# Phyloseq obj for alpha diversity for alcohol
```{r alcohol vs non alcohol}
# counts
counts_alcohol = as.matrix(counts_alcohol)
alcohol_otu = otu_table(counts_alcohol, taxa_are_rows = T)
# metadata set
metadata_alcohol_sample = sample_data(metadata_alcohol)
# taxonomy dataset
taxonomy_alcohol_data = taxonomy_alcohol |>
  separate(col = Taxon,
           into = c('Domain', 'Phylum', 'Class', 'Order',
                     'Family', 'Genus', 'Species'),
            sep=';', fill = 'right') |>
  select(-Confidence) |>
  as.matrix()
taxonomy_alcohol_data = tax_table(taxonomy_alcohol_data)

# phyloseq combine

# intersect(taxa_names(alcohol_otu), taxa_names(taxonomy_alcohol_altered))
ps_alcohol = phyloseq(alcohol_otu, metadata_alcohol_sample, taxonomy_alcohol_data)
ps_alcohol
```
# Alpha shannon diversity for alcohol vs non alcohol
## Panel A
```{r}
# ps = readRDS('C:/MICB_305/Do_Thea_Assignment5/phyloseq_taxonomy.rds')
# Histogram is for the overview of ps_alcohol
hist(sample_sums(ps_alcohol))
# This is similar to what we have seen in qiime, it's just that it's clearer
rarecurve(t(data.frame(ps_alcohol@otu_table)),
          step=500,
          label = FALSE)
# Sampling depth is sample.size
# rngseed is random seed (set.seed)
psrare_alcohol = ps_alcohol |>
  rarefy_even_depth(sample.size = 9000, rngseed=123)

table(sample_sums(psrare_alcohol)) # ensure remaining samples have the same sampling depth as the chosen one above

# Calculate Shannon alpha diversity
# This is for a general observation of Shannon diversity plot
shannon = plot_richness(psrare_alcohol, x = "alcohol_intake_status", measures = "Shannon")

set.seed(123)
plot_alcohol = plot_richness(psrare_alcohol, x = "alcohol_intake_status",
                     measures = "Shannon",
                     color = "alcohol_intake_status",
                     )
plot_alcohol
# Now make a publication-ready plot
pdata_alcohol = plot_alcohol$data
str(pdata_alcohol)

shannon_alcohol_plot <- pdata_alcohol |>
  ggplot(aes(alcohol_intake_status, value, fill = alcohol_intake_status)) +
  geom_boxplot(outlier.shape = NA, width = 0.6) +
  geom_jitter(height = 0.2, width = 0.2, size = 2, alpha = 0.7) +
  theme_classic(base_size = 14) +
  labs(fill = "alcohol_intake_status") + 
  scale_fill_manual(values = c("Low" = "#CC79A7",
                               "High" = "#A6CEE3")) +
  theme(legend.position = "none") +
  labs(title = "Shannon Alpha Diversity between WT and Susceptible Groups",
       x = "alcohol_intake_status",
       y = "Shannon Diversity Index") +
  stat_compare_means(comparisons = list(c("Low", "High")),
                     method = "wilcox.test",
                     label = "p.format",
                     label.y = max(pdata_alcohol$value) + 0.2)
shannon_alcohol_plot
```
```{r Statistical test}
# Rerun statistical test for alcohol to make sure it is correctly plotted
kruskal.test(value ~ alcohol_intake_status, data = pdata_alcohol)
# p-value = 0.2248, which indicates non-significant difference between low and high alcohol intake groups
```

# Phyloseq object for IBD
```{r}
# counts
counts_ibd = as.matrix(counts_ibd)
ibd_otu = otu_table(counts_ibd, taxa_are_rows = T)
# metadata set
metadata_ibd_sample = sample_data(metadata_ibd)
# taxonomy dataset
taxonomy_ibd_data = taxonomy_ibd |>
  separate(col = Taxon,
           into = c('Domain', 'Phylum', 'Class', 'Order',
                     'Family', 'Genus', 'Species'),
            sep=';', fill = 'right') |>
  select(-Confidence) |>
  as.matrix()
taxonomy_ibd_data = tax_table(taxonomy_ibd_data)

# phyloseq combine

# intersect(taxa_names(alcohol_otu), taxa_names(taxonomy_alcohol_altered))
ps_ibd = phyloseq(ibd_otu, metadata_ibd_sample, taxonomy_ibd_data)
ps_ibd
```
# Alpha shannon diversity for IBD
## Panel B
```{r}
# Histogram is for the overview of ps_alcohol
hist(sample_sums(ps_ibd))
# This is similar to what we have seen in qiime, it's just that it's clearer
rarecurve(t(data.frame(ps_ibd@otu_table)),
          step=1000,
          label = FALSE)
# Sampling depth is sample.size
# rngseed is random seed (set.seed)
psrare_ibd = ps_ibd |>
  rarefy_even_depth(sample.size = 4167, rngseed=123)

table(sample_sums(psrare_ibd)) # ensure remaining samples have the same sampling depth as the chosen one above

# Calculate Shannon alpha diversity
# This is for a general observation of Shannon diversity plot
shannon_ibd = plot_richness(psrare_ibd, x = "Condition", measures = "Shannon")
shannon_ibd

set.seed(123)
plot_ibd = plot_richness(psrare_ibd, x = "Condition",
                     measures = "Shannon",
                     color = "Condition",
                     )
plot_ibd
# Now make a publication-ready plot
pdata_ibd = plot_ibd$data
str(pdata_ibd)

comparisons = list(c("Healthy", "Ulcerative Colitis"), c("Ulcerative Colitis", "Crohn's Disease"), c("Healthy", "Crohn's Disease"))
y_positions = c(max(pdata_ibd$value + 0.1),
                max(pdata_ibd$value + 0.2),
                max(pdata_ibd$value + 0.3))

shannon_ibd_plot <- pdata_ibd |>
  ggplot(aes(Condition, value, fill = Condition)) +
  geom_boxplot(outlier.shape = NA, width = 0.6) +
  geom_jitter(height = 0.2, width = 0.2, size = 2, alpha = 0.7) +
  theme_classic(base_size = 14) +
  labs(fill = "Condition") + 
  scale_fill_manual(values = c("Crohn's Disease" = "#CC79A7",
                               "Ulcerative Colitis" = "#A6CEE3",
                               "Healthy" = "#CAB2D6")) +
  theme(legend.position = "none") +
  labs(title = "Shannon Alpha Diversity between UC, CD, and Healthy Groups",
       x = "Condition",
       y = "Shannon Diversity Index") +
  stat_compare_means(comparisons = comparisons,
                     method = "wilcox.test",
                     label = "p.format",
                     y.position = y_positions)
shannon_ibd_plot
```
```{r Statistical test}
# Kruskal-Wallis test
kruskal.test(value ~ Condition, data = pdata_ibd)
# At least 1 group is significantly different
dunn.test(pdata_ibd$value, pdata_ibd$Condition, method = "BH")
```

# Jaccard for alcohol
## Panel C
```{r}
# Jaccard considers richness only (no evenness or phylogenetic distances)
ps_jaccard_alcohol = phyloseq::distance(psrare_alcohol, method = "jaccard", binary = TRUE)
View(as.matrix(ps_jaccard_alcohol)) # Every pairwise comparison included

# MDS scaling
set.seed(123)
mds_alcohol = metaMDS(ps_jaccard_alcohol)

# Extract data
mds_data_alcohol = mds_alcohol$points %>%
  as.data.frame %>%
  merge(sample_data(psrare_alcohol), by='row.names', sort=F)
head(mds_data_alcohol)

# Plot alcohol Jaccard diversity (Not Venn Diagram)
alcohol_plot = mds_data_alcohol %>%
  ggplot(aes(MDS1,MDS2,color = alcohol_intake_status)) +
  geom_point(size = 3) +
  stat_ellipse() +
  theme_classic(base_size=18) +
  scale_colour_manual(values = c("Low" = "#A6CEE3",
                                 "High" = "#CAB2D6")) +
  labs(title = "Jaccard Beta-Diversity between Alcohol Intake Groups",
       colour = "Alcohol Intake Status")
alcohol_plot

# Venn Diagram?
# This is to separate OTUs of low and high alcohol intake
# Extract sample names by group
low_alcohol = sample_names(ps_alcohol)[sample_data(ps_alcohol)$alcohol_intake_status == "Low"]
high_alcohol = sample_names(ps_alcohol)[sample_data(ps_alcohol)$alcohol_intake_status == "High"]

# Get OTUs present in each group
otu_low = unique(unlist(taxa_names(prune_samples(low_alcohol, ps_alcohol))))
otu_high = unique(unlist(taxa_names(prune_samples(high_alcohol, ps_alcohol))))

# Create list
otu_list <- list(Low = otu_low, High = otu_high)

# Draw Venn Diagram using ggplot
venn_diagram = ggVennDiagram(otu_list,
                             label_alpha = 0,
                             set_size = 5) +
  scale_fill_gradient(low = "#A6CEE3", high = "#CAB2D6") + 
  labs(title = "Venn Diagram")
venn_diagram
```

```{r Statistical Test}
all(rownames(pdata_alcohol) == rownames(as.matrix(ps_jaccard_alcohol)))
pdata_alcohol_clean = select(pdata_alcohol, -se)

adonis2(ps_jaccard_alcohol ~ alcohol_intake_status, data = pdata_alcohol_clean)
# p-value = 0.488, indicating no significant difference
# Since there are only 2 groups for alcohol intake, I did not use pairwise permanova here
```

# Jaccard for IBD
## Pannel D
```{r}
ps_jaccard_ibd = phyloseq::distance(psrare_ibd, method = "jaccard", binary = TRUE)
View(as.matrix(ps_jaccard_ibd)) # Every pairwise comparison included

# MDS scaling
set.seed(123)
mds_ibd = metaMDS(ps_jaccard_ibd)

# Extract data
mds_data_ibd = mds_ibd$points %>%
  as.data.frame %>%
  merge(sample_data(psrare_ibd), by='row.names', sort=F)
head(mds_data_ibd)

# Plot IBD Jaccard diversity 
ibd_plot = mds_data_ibd %>%
  ggplot(aes(MDS1,MDS2,color = Condition)) +
  geom_point(size = 3) +
  stat_ellipse() +
  theme_classic(base_size=18) +
  scale_colour_manual(values = c("Ulcerative Colitis" = "#A6CEE3",
                                 "Crohn's Disease" = "#CAB2D6",
                                 "Healthy" = "#CC79A7")) +
  labs(title = "Jaccard Beta-Diversity between Healthy and IBD Patients",
       colour = "Condition")
ibd_plot
```
```{r Statistical Test}
adonis2(ps_jaccard_ibd ~ Condition, data = pdata_ibd)
# Overall p-value between 3 groups is 0.001, which indicates significant difference between all 3 groups

# Now doing pairwise PERMANOVA to see if there is significant difference between conditions
pairwise.adonis(ps_jaccard_ibd, factors = pdata_ibd$Condition, 
                sim.method = "jaccard", p.adjust.m = "BH") # in dunn.test I used BH for method so maybe I should use it consistently?
# Look at p.adjusted for p-values between different conditions
# Ulcerative Colitis vs Crohn's Disease: p-value = 0.0020 (significant difference)
# Ulcerative Colitis vs Healthy: 0.0015 (significant difference)
# Crohn's Disease vs Healthy: 0.0015 (significant difference)
```
# VICTORY!!
